# 第04课：条件与判断

本笔记本是课堂讲义。每个知识点包含：理论知识、案例代码、讲解、易错点与练习。综合练习 P1 使用同目录的 [mobile_game_info.csv](mobile_game_info.csv)。课后独立练习见 [chapter04_条件与判断_课后练习.ipynb](chapter04_条件与判断_课后练习.ipynb)，数据为 [googleplaystore.csv](googleplaystore.csv)。

## 学习目标

1. 说明比较运算的结果是布尔值 `True` / `False`，并能区分文件里的文本 `"TRUE"` 与布尔 `True`。
2. 在 `for` 循环中使用 `if` 决定某一行要不要留下，解释缩进范围。
3. 使用 `!=`、`and`、`or` 与括号写出筛选规则，并能口头说明留下谁、扔掉谁。
4. 对筛选后的记录转换评分、累加并求平均，分母只数被留下的记录。
5. 使用 `if` / `elif` / `else` 给评分分档，并检查三类计数之和等于参与分档的记录数。

## 学习知识点

| 比较与布尔 | 筛选一行 | 组合与分档 |
| --- | --- | --- |
| `== != > >= < <=` | `if` 与缩进 | `and` 同时满足 |
| `True` / `False` | 循环里决定去留 | `or` 满足其一 |
| 文本 `"TRUE"` | 子集计数与平均分 | `elif` 分档与守恒 |

## 基础回顾与案例提问

第3课已经把表读成 `body`。下面这条记录摘自原文件第一行，字段顺序与完整七列一致：

`record = ["206369", "人格解体", "8.8", "角色扮演", "厂商: 喵法自然工作室", "71069", "TRUE"]`

1. **R.1** `record[6]` 的类型是什么？`record[6] == True` 和 `record[6] == "TRUE"` 哪一个为真？先预测再写表达式验证。
2. **R.2** 不转换类型，直接写 `record[2] >= 8.5`，比较的是什么？要按数值高低判断，应先做什么？
3. **R.3** 若只统计安卓记录的平均分，分母应是 `len(body)` 还是安卓记录数？为什么？

**作答：** 预测：____；依据：____；验证后说明：____。

所有游戏案例均取自 [mobile_game_info.csv](mobile_game_info.csv)。先用少量摘录记录讲语法，再读取同一文件。使用 Python 3；只需标准库 `csv`。CSV 与 Notebook 放在同一文件夹，从该文件夹启动内核，按顺序运行案例。本课 CSV 可能带有 UTF-8 BOM，打开文件时使用 `encoding="utf-8-sig"`。本课不讲列表推导式、三元表达式、`break` / `continue` 或自定义函数。


In [ ]:
# R.1–R.3: Write and verify your predictions here.


## 1. 比较运算与布尔值

### 理论知识

**比较运算用来判断两个值的关系，结果只有两种：`True` 或 `False`。** 这种结果的类型是 `bool`（布尔）。

| 运算 | 含义 |
| --- | --- |
| `==` | 相等 |
| `!=` | 不相等 |
| `>` `>=` | 大于、大于或等于 |
| `<` `<=` | 小于、小于或等于 |

比较不会修改原来的值；它只产生一个判断结果。赋值 `=` 与比较 `==` 不是同一件事：前者给名字，后者问是否相等。

两个运算元的类型会影响比较含义。数字按数值大小比较；字符串按文本比较，不是按数值大小。看起来像数字的字段，若仍是 `str`，就还不是数值比较。

### 案例：一条记录上的判断结果


In [1]:
platform_text = "TRUE"
score_text = "8.8"
print(platform_text == "TRUE")
print(platform_text == True)
print(type(platform_text == "TRUE"))
print(score_text >= "8.5")
print(float(score_text) >= 8.5)


True
False
<class 'bool'>
True
True


### 讲解

第一个比较为 `True`，因为两边都是文本 `"TRUE"`。第二个为 `False`：左边是字符串，右边是布尔值，二者不相等。`type(...)` 显示比较结果的类型是 `bool`。

`"8.8" >= "8.5"` 是字符串比较，碰巧也得到 `True`，但不能当成评分规则。`float(score_text) >= 8.5` 才是数值比较。后文凡是按评分高低筛选，都先 `float`。

### 易错点与练习

把 `==` 写成 `=` 不是“比错了”，而是改成了赋值，常常直接语法错误，或把条件位置变成意外赋值。排查时先看是不是少写了一个等号。

1. **K1.1** 对 `game_type = "策略"`，写出 `game_type == "角色扮演"` 和 `game_type != "角色扮演"` 的结果，并说明二者为什么不会同时为真。
2. **K1.2** 预测 `"9" > "10"` 与 `float("9") > float("10")`。先写预测再运行，解释为什么字符串比较不能代替评分比较。

**作答：** 两个布尔结果：____；字符串比较与数值比较的差异：____。


In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.


## 2. if 语句：条件成立才执行

### 理论知识与语法

**`if` 根据一个布尔值决定要不要执行某段代码。** 基本形式：

```python
if condition:
    statement_1
    statement_2
statement_after
```

- `if` 和冒号是语法的一部分；`condition` 写比较或已经得到的布尔值。
- 冒号后面缩进的语句组成 `if` 体；本课统一四个空格。
- `condition` 为 `True` 时执行缩进语句；为 `False` 时跳过它们。
- `statement_after` 没有缩进，无论条件真假都会执行。

`if` 本身不会自动遍历表格。它一次只判断当前这个条件。要检查每一行，需要和 `for` 一起用：`for` 负责取出下一行，`if` 负责这一行要不要处理。

### 案例：只在平台字段匹配时打印


In [2]:
one_game = ["人格解体", "8.8", "角色扮演", "TRUE"]
if one_game[3] == "TRUE":
    print("Keep:", one_game[0])
print("Finished one record")


Keep: 人格解体
Finished one record


### 讲解

`one_game[3] == "TRUE"` 为 `True`，因此打印名称。最后一句没有缩进，判断结束后一定会执行。若把平台改成 `"FALSE"`，中间的 `print` 不会运行，但结束提示仍会打印。

下面用五条摘录记录跟踪：哪些行会进入 `if` 体，哪些行只经过循环、不打印名称。简化表的字段为 title、score、game_type、isAndroid，下标依次是 0、1、2、3。这与完整七列表头不同，后文读取文件时会改回完整下标。


In [3]:
sample = [
    ["人格解体", "8.8", "角色扮演", "TRUE"],
    ["一念逍遥", "6.5", "角色扮演", "TRUE"],
    ["明日方舟", "8.2", "策略", "TRUE"],
    ["另一个伊甸 : 超越时空的猫", "9", "角色扮演", "FALSE"],
    ["偶像梦幻祭2", "8.2", "音乐", "FALSE"],
]
for row in sample:
    if row[3] == "TRUE":
        print(row[0])
print("Finished sample")


人格解体
一念逍遥
明日方舟
Finished sample


### 易错点与练习

循环变量 `row` 每轮是一整行。条件里应写 `row[3]`，不要写 `sample[3]`（那是第四条记录）。`if` 体只包含缩进语句；把需要“每条匹配记录都做”的 `print` 移出 `if`，就只看到最后一次判断之后的结果。

1. **K2.1** 对 `sample` 写出：仅当类型为 `"角色扮演"` 时打印名称。先列出哪几轮会打印，再运行核对。
2. **K2.2** 下方故障代码想只打印安卓游戏名称，实际会怎样？预测输出并修正缩进。

```python
# Deliberate logic fault; keep it outside normal execution cells.
for row in sample:
    if row[3] == "TRUE":
        name = row[0]
    print(name)
```

**作答：** 会打印的名称：____；故障输出预测/原因/修正：____。


In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.


## 3. 读取课程文件并确认列位置

### 理论知识

筛选之前必须先确认**当前这张表**的列顺序。列下标不属于“评分”或“安卓”这两个概念本身，而属于眼前这份文件的表头。

本课完整文件的列如下：

| 索引 | 列名 | 本课用法 |
| --- | --- | --- |
| 0 | id | 标识 |
| 1 | title | 游戏名称 |
| 2 | score | 评分，计算前转 `float` |
| 3 | game_type | 类型，例如角色扮演、策略 |
| 4 | pulisher | 厂商信息，保留源拼写；本课筛选不用 |
| 5 | download | 下载量；本课筛选不用 |
| 6 | isAndroid | 平台标记，文本 `TRUE` / `FALSE` |

简化四字段表里，评分下标是 1、类型是 2、平台是 3；完整文件里评分是 2、类型是 3、平台是 6。从摘录练习切到全表时，只在这一处改下标。

文件中存在重复标题，本课不去重：统计的是记录，不是唯一游戏数量。

### 案例：读入 header 与 body


In [4]:
import csv

with open("mobile_game_info.csv", "r", encoding="utf-8-sig", newline="") as file:
    rows = list(csv.reader(file))
header = rows[0]
body = rows[1:]
print("Column names:", header)
print("Score index:", header.index("score"))
print("Type index:", header.index("game_type"))
print("Platform index:", header.index("isAndroid"))
print("First data record:", body[0])
print("Record count:", len(body))


Column names: ['id', 'title', 'score', 'game_type', 'pulisher', 'download', 'isAndroid']
Score index: 2
Type index: 3
Platform index: 6
First data record: ['206369', '人格解体', '8.8', '角色扮演', '厂商:\xa0喵法自然工作室', '71069', 'TRUE']
Record count: 266


### 讲解与易错点

`header.index("score")` 用列名核对下标，避免凭记忆写错。第一行数据的平台字段是文本 `'TRUE'`，评分字段是文本 `'8.8'`。`len(body)` 是数据记录数，不含表头。

打开失败时先检查工作目录和文件名；中文乱码时检查是否漏了 `utf-8-sig`。不要把表头行送进后面的平台比较。

### 练习

1. **K3.1** 打印 `body[2]` 的 title、score、game_type、isAndroid。说明这四个值分别用了哪些下标，以及它们和 `sample[2]` 的下标有何不同。
2. **K3.2** `len(header)`、`len(body)`、`len(rows)` 各表示什么？若用 `len(rows)` 当筛选前的记录总数，会多算什么？

**作答：** 四个字段与下标：____；三个长度的含义：____。


In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.


## 4. 在循环中筛选：留下安卓记录

### 理论知识与语法

筛选一张表，标准写法是：**遍历每一行，用 `if` 判断是否留下，把留下的行放进新列表。**

```python
selected = []
for row in body:
    if condition:
        selected.append(row)
```

`append` 把当前行加到 `selected` 末尾。它不是新的数据类型，只是列表“在末尾增加一个元素”的方法。本课用它收集筛选结果，不展开其它列表方法。

- `selected` 必须在循环前建成空列表，循环内不要重新写成 `[]`。
- `condition` 为 `False` 的行不会 `append`，也就不会进入后续平均分。
- `len(selected)` 是留下的记录数，作为子集平均分的分母候选。

### 案例：收集 isAndroid 为文本 TRUE 的记录


In [5]:
android_rows = []
for row in body:
    if row[6] == "TRUE":
        android_rows.append(row)
print("Android count:", len(android_rows))
print("First android title:", android_rows[0][1])
print("Original record count:", len(body))


Android count: 150
First android title: 人格解体
Original record count: 266


### 讲解

全表 266 条记录里，平台字段等于 `"TRUE"` 的被放进 `android_rows`。`body` 本身没有被删改；筛选得到的是另一份列表。打印 `len(android_rows)` 与 `len(body)`，可以看见子集和全表不是同一个范围。

只想看前几条匹配结果时，仍先按全表规则筛选，再对结果切片，不要把 `body[:5]` 误当成“全表安卓”。

### 易错点与练习

`if row[6] = "TRUE"` 少写等号。`if row == "TRUE"` 拿整行去和字符串比，几乎总是 `False`。循环内写 `android_rows = []` 会每轮清空已收集的行。

1. **K4.1** 用同样结构收集类型为 `"策略"` 的记录，打印数量，并打印其中第一条的 title。先根据表头确认类型下标。
2. **K4.2** 对 `sample` 收集平台为 `"FALSE"` 的行。先手写应留下哪些名称，再运行比较。

**作答：** 策略记录的下标与数量：____；sample 中非安卓名称：____。


In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.


## 5. 文本 "TRUE" 与布尔 True

### 理论知识

CSV 读进来的字段默认是字符串。本文件的 `isAndroid` 取值是文本 `"TRUE"` 和 `"FALSE"`，**不是**布尔值 `True` 和 `False`。

判断平台时应写：

```python
row[6] == "TRUE"
```

不要写 `row[6] == True`，也不要写 `if row[6]:`。非空字符串在 `if` 里会被当成真，那样 `"FALSE"` 也会被留下，因为 `"FALSE"` 仍是非空文本。

`type(row[6])` 应显示 `str`。是否安卓，以文件里的真实文本为准。

### 案例：同一条记录的三种判断


In [6]:
platform_field = body[0][6]
print("Raw value:", platform_field, type(platform_field))
print("Equal to text TRUE:", platform_field == "TRUE")
print("Equal to boolean True:", platform_field == True)
if platform_field:
    print("Non-empty text is treated as true in if")
false_text = "FALSE"
print("FALSE text in if:", bool(false_text))
print("FALSE text == TRUE:", false_text == "TRUE")


Raw value: TRUE <class 'str'>
Equal to text TRUE: True
Equal to boolean True: False
Non-empty text is treated as true in if
FALSE text in if: True
FALSE text == TRUE: False


### 讲解与易错点

第一行平台文本是 `'TRUE'`，与字符串比较为真，与布尔 `True` 比较为假。`if platform_field:` 只问“这串文本空不空”，因此 `"TRUE"` 和 `"FALSE"` 都会走进 `if` 体。筛选安卓必须用 `== "TRUE"`。

`bool("FALSE")` 为 `True` 常让人意外。本课不把 `bool(...)` 当作平台转换函数。

### 练习

1. **K5.1** 打印 `body[0][6]` 与 `True`、`"TRUE"`、`"True"` 三个比较结果，说明为什么大小写不同的文本也不相等。
2. **K5.2** 若有人用 `if row[6]:` 收集“安卓行”，对 `sample` 会留下几行？写出应改成的条件。

**作答：** 三个比较：____；错误条件会留下的行数及修正：____。


In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.


## 6. != 筛选非安卓记录

### 理论知识与语法

`!=` 表示不相等。找非安卓记录时，与文件中的安卓标记相反即可：

```python
row[6] != "TRUE"
```

在本文件中，平台字段只有 `"TRUE"` 和 `"FALSE"`，因此 `row[6] != "TRUE"` 与 `row[6] == "FALSE"` 选出同一批记录。书写时仍以“不是安卓标记”这个规则为准，先核对表里实际出现过哪些取值。

若分类完整且互不重叠，则：

```text
安卓记录数 + 非安卓记录数 = 数据记录数
```

这个等式用来检查有没有漏行，不是新的平均分公式。

### 案例：对照安卓与非安卓数量


In [7]:
not_android_rows = []
for row in body:
    if row[6] != "TRUE":
        not_android_rows.append(row)
print("Android:", len(android_rows))
print("Not android:", len(not_android_rows))
print("Sum:", len(android_rows) + len(not_android_rows))
print("Body:", len(body))


Android: 150
Not android: 116
Sum: 266
Body: 266


### 讲解

两份列表数量相加应等于 `len(body)`。不相等时，优先检查条件是否写成了 `== True`、是否把表头算进去，或是否在循环里清空了列表。

`sample` 里非安卓是后两条：另一个伊甸与偶像梦幻祭2。完整文件不要用手算 266 条来验收，用程序计数并核对守恒即可。

### 易错点与练习

`row[6] != True` 几乎对每一行都为真，因为字段是字符串。守恒失败时不要先改平均分公式，先打印两个 `len`。

1. **K6.1** 对 `sample` 分别收集 `!= "TRUE"` 与 `== "FALSE"` 的名称，说明这两份名称为何相同。
2. **K6.2** 已知某次运行安卓 150、非安卓 110、`len(body)` 为 266。指出守恒是否成立，可能漏掉的是哪一类检查。

**作答：** 两份名称：____；守恒判断：____。


In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.


## 7. 子集平均分：只累计被留下的评分

### 理论知识与公式

筛选后的平均分，分子和分母必须来自**同一批被留下的记录**：

```text
子集平均分 = 留下记录的评分总和 / 留下记录的条数
```

循环里同时完成两件事：判断是否留下，以及（若留下）把评分转成浮点数后累加。计数器也只在条件成立时加 1。

```python
total = 0.0
count = 0
for row in selected_rows:
    score = float(row[score_index])
    total = total + score
    count = count + 1
average = total / count
```

也可以直接在全表循环里写 `if`，不必先建 `selected_rows`；两种写法的口径必须一致。`count` 为 0 时不能相除。本课演示中的子集均非空。

### 案例：角色扮演记录的平均分


In [8]:
rpg_total = 0.0
rpg_count = 0
for row in body:
    if row[3] == "角色扮演":
        rpg_total = rpg_total + float(row[2])
        rpg_count = rpg_count + 1
rpg_average = rpg_total / rpg_count
print("RPG count:", rpg_count)
print("RPG total:", rpg_total)
print("RPG mean:", rpg_average)


RPG count: 67
RPG total: 514.1000000000001
RPG mean: 7.673134328358211


### 代码讲解

只有 `game_type` 为 `"角色扮演"` 的行参与累加。分母是 `rpg_count`，不是 `len(body)`，也不是字段个数 7。全表平均分会掩盖“只看角色扮演”的结果，这正是本课要做筛选的原因。浮点输出可能出现末位尾差，不改变这些记录被计入的口径。

对 `sample` 可先手算：角色扮演为 8.8、6.5、9，总和 24.3，条数 3，平均 8.1。手算只核对手头这三行，不要拿去和全表角色扮演总分比较。

### 易错点与练习

忘记 `float` 时，`total` 可能变成字符串拼接，或直接 `TypeError`。用 `len(body)` 当子集分母，会把被扔掉的行也算进平均。`count` 初始化若放进 `if` 里面且每次重置，计数会错。

1. **K7.1** 对 `sample` 计算策略记录的平均分。先手算再写循环，分母只数策略行。
2. **K7.2** 下方代码已按角色扮演累加总分，却除以 `len(body)`。指出分母混入了什么，并写出修正。

```python
# Deliberate logic fault.
print(rpg_total / len(body))
```

**作答：** 手算/程序对照：____；错误分母及修正：____。


In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.


## 8. and：同时满足

### 理论知识与语法

`and` 连接两个条件，**左右都为 `True` 时，整个表达式才为 `True`**。

```python
condition_a and condition_b
```

| 左 | 右 | `and` |
| --- | --- | --- |
| False | False | False |
| False | True | False |
| True | False | False |
| True | True | True |

它不是把两个筛选结果的行数相加。它要求同一条记录同时符合两条规则，例如“类型是角色扮演，并且平台是安卓”。

### 案例：角色扮演且安卓


In [9]:
print("Sample traces:")
for row in sample:
    is_rpg = row[2] == "角色扮演"
    is_android = row[3] == "TRUE"
    print(row[0], is_rpg, is_android, is_rpg and is_android)

rpg_android = []
for row in body:
    if row[3] == "角色扮演" and row[6] == "TRUE":
        rpg_android.append(row)
print("Full-file RPG and android:", len(rpg_android))


Sample traces:
人格解体 True True True
一念逍遥 True True True
明日方舟 False True False
另一个伊甸 : 超越时空的猫 True False False
偶像梦幻祭2 False False False
Full-file RPG and android: 31


### 讲解

`sample` 中同时为真的是人格解体与一念逍遥；另一个伊甸是角色扮演但非安卓，明日方舟是安卓但类型为策略。完整文件使用下标 3 与 6。不要把 `sample` 的 2 条拿去和全表数量比较。

### 易错点与练习

写成两个独立的 `if` 且都 `append` 到同一列表，一条记录可能被加两次，或口径变成“满足任一条件”。`and` 的两边应是布尔条件，不要写成 `row[3] and row[6] == "TRUE"` 这种只比较了右侧的形式。

1. **K8.1** 对 `sample` 收集“策略且安卓”的名称。先手写再运行。
2. **K8.2** 判断 `row[3] == "角色扮演" and "策略"` 是否能表示“类型为角色扮演或策略”？先预测这个表达式在人格解体那一行的真假，再解释它实际在问什么。

**作答：** 名称：____；错误表达式的含义：____。


In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.


## 9. or：满足其一

### 理论知识与语法

`or` 连接两个条件，**左右至少有一个为 `True` 时，整个表达式为 `True`**。两个都为真也会留下，不是“只能选一边”。

```python
condition_a or condition_b
```

| 左 | 右 | `or` |
| --- | --- | --- |
| False | False | False |
| False | True | True |
| True | False | True |
| True | True | True |

日常口语里的“或者”有时被理解成互斥。Python 的 `or` 不是互斥：角色扮演并且同时也能被其它规则说中的记录，仍然只是一条记录、会被留下一次。

源实验材料里若把 `or` 说成“同时满足”，以课堂规则为准：**同时满足是 `and`，满足其一是 `or`。**

### 案例：角色扮演或策略


In [10]:
print("Sample traces:")
for row in sample:
    is_rpg = row[2] == "角色扮演"
    is_strategy = row[2] == "策略"
    print(row[0], is_rpg, is_strategy, is_rpg or is_strategy)

rpg_or_strategy = []
for row in body:
    if row[3] == "角色扮演" or row[3] == "策略":
        rpg_or_strategy.append(row)
print("Full-file RPG or strategy:", len(rpg_or_strategy))


Sample traces:
人格解体 True False True
一念逍遥 True False True
明日方舟 False True True
另一个伊甸 : 超越时空的猫 True False True
偶像梦幻祭2 False False False
Full-file RPG or strategy: 93


### 讲解

`sample` 中偶像梦幻祭2 的类型是音乐，两个条件都为假，被排除；其余四条至少满足一侧。人格解体只满足左侧，明日方舟只满足右侧，它们都会留下。这就是“满足其一”，不是“两个条件要一起成立”。

完整文件里类型只有一个取值，一条记录不会既是角色扮演又是策略；即便如此，`or` 的定义仍然包含“两侧都真”的那一行。

### 易错点与练习

`row[3] == "角色扮演" or "策略"` 的右侧是非空字符串，恒为真，几乎会留下所有行。两个条件都要写完整比较。

1. **K9.1** 对 `sample` 收集类型为 `"音乐"` 或 `"策略"` 的名称，说明哪些只满足一侧。
2. **K9.2** 预测 `row[3] == "角色扮演" or "策略"` 对 `sample` 每一行的真假，并写出修正后的 `or` 条件。

**作答：** 名称与哪一侧为真：____；故障原因与修正：____。


In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.


## 10. 组合条件与括号

### 理论知识

实际筛选常常是两类规则叠在一起，例如：**类型为角色扮演或策略，并且不是安卓。**

写成：

```python
row[6] != "TRUE" and (row[3] == "角色扮演" or row[3] == "策略")
```

括号是必要的。在 Python 中 `and` 的结合优先于 `or`：没有括号时，`A and B or C` 相当于 `(A and B) or C`，口径就变了。先用中文说出规则，再给“且 / 或”两边加上完整比较，最后用括号把“或”包在一起。

口头验收时要说得出：留下非安卓，且类型属于角色扮演、策略二者之一；安卓记录全部扔掉；音乐等其它类型即使非安卓也扔掉。

### 案例：先在 sample 上跟踪，再写出全表代码结构


In [11]:
print("Sample traces:")
for row in sample:
    not_android = row[3] != "TRUE"
    rpg_or_strategy = row[2] == "角色扮演" or row[2] == "策略"
    keep = not_android and rpg_or_strategy
    print(row[0], not_android, rpg_or_strategy, keep)


Sample traces:
人格解体 False True False
一念逍遥 False True False
明日方舟 False True False
另一个伊甸 : 超越时空的猫 True True True
偶像梦幻祭2 True False False


### 讲解

`sample` 里只有“另一个伊甸 : 超越时空的猫”同时满足非安卓以及（角色扮演或策略）。偶像梦幻祭2 非安卓但类型是音乐；人格解体等三条是目标类型但是安卓。

全表不要在这一节直接公布最终平均分。把同一规则换到完整下标，放进综合练习由自己算。比较时只能拿同一规则、同一批记录对照。

### 易错点与练习

漏掉括号会把规则改成另一种筛选。`and` 与 `or` 写反时，行数往往会明显变大或变小，应先用 `sample` 跟踪，再相信全表数字。

1. **K10.1** 对 `sample` 收集“安卓且（角色扮演或策略）”的名称。先手写再运行。
2. **K10.2** 比较下面两个条件在 `sample` 上各留下谁，说明括号改变了什么。

```python
# Deliberate contrast; keep both in Markdown.
row[3] != "TRUE" and row[2] == "角色扮演" or row[2] == "策略"
row[3] != "TRUE" and (row[2] == "角色扮演" or row[2] == "策略")
```

**作答：** K10.1 名称：____；两个条件留下的行及差异：____。


In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.


## 11. 数值比较：先转换再判断高低

### 理论知识

`>=`、`>`、`<=`、`<` 按数值分档前，评分必须已经是 `float`。文本 `"8.8"` 与数字 `8.5` 不能直接比大小，会 `TypeError`；两个字符串相比又不是评分规则。

正确顺序：

```python
score = float(row[2])
if score >= 8.5:
    ...
```

阈值本身应写成数字 `8.5`，不要写成 `"8.5"`。本课高分档阈值是 **8.5**，中分档阈值是 **7.0**，与验收脚本一致。

### 案例：在 sample 上看转换后的分档判断


In [12]:
for row in sample:
    score = float(row[1])
    print(row[0], row[1], type(row[1]), score, score >= 8.5)


人格解体 8.8 <class 'str'> 8.8 True
一念逍遥 6.5 <class 'str'> 6.5 False
明日方舟 8.2 <class 'str'> 8.2 False
另一个伊甸 : 超越时空的猫 9 <class 'str'> 9.0 True
偶像梦幻祭2 8.2 <class 'str'> 8.2 False


### 讲解

简化表评分在下标 1；完整文件在下标 2。人格解体 8.8、另一个伊甸 9 达到高分阈值；明日方舟 8.2 与偶像梦幻祭2 8.2 未达到 8.5；一念逍遥 6.5 更低。`print` 里的布尔值只是判断结果，还没有计数。

### 易错点与练习

`row[2] >= 8.5` 在完整文件上比较的是字符串和数字。`float(row[2]) >= "8.5"` 则把阈值又写成了字符串。分档前先打印 `type(score)` 确认。

1. **K11.1** 对 `sample` 打印每条是否 `score >= 7.0`。先手写五条的 True/False，再运行。
2. **K11.2** 预测 `float(body[0][2]) >= 8.5` 与 `body[0][2] >= 8.5` 哪一个能按数值规则工作，哪一个会出错或没有数值含义。不要把故障语句留在正常运行流程里。

**作答：** 五条判断：____；两种写法的区别：____。


In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.


## 12. if / elif / else：分档与计数守恒

### 理论知识与语法

一条记录在同一套分档规则里只应落入一档。`elif` 表示“上面都不成立时，再问下一个条件”；`else` 承接剩余情况。

```python
if score >= 8.5:
    high = high + 1
elif score >= 7.0:
    mid = mid + 1
else:
    low = low + 1
```

顺序有含义：先问高分，再问中分。若先写 `>= 7.0`，高分记录也会先进中档，后面的 `>= 8.5` 再也轮不到。三个分支互斥，因此：

```text
high + mid + low = 参与分档的记录数
```

对全表分档时，右侧是 `len(body)`。不要用三个独立的 `if`（不用 `elif`）去加三个计数，同一条高分记录可能被多个条件重复统计。

### 案例：sample 五条记录分档


In [13]:
high = 0
mid = 0
low = 0
for row in sample:
    score = float(row[1])
    if score >= 8.5:
        high = high + 1
    elif score >= 7.0:
        mid = mid + 1
    else:
        low = low + 1
print("High / mid / low:", high, mid, low)
print("Sum:", high + mid + low)
print("Sample size:", len(sample))


High / mid / low: 2 2 1
Sum: 5
Sample size: 5


### 讲解

8.8 与 9 记入高分，两条 8.2 记入中分，6.5 记入低分。2 + 2 + 1 = 5，与 `len(sample)` 一致。全表分档把 `row[1]` 换成 `row[2]`，并把循环对象换成 `body`；具体三个计数留给综合练习独立完成。

### 易错点与练习

三个都写成 `if` 会破坏互斥。漏掉 `else` 时，低于 7.0 的记录不进入任何档，守恒失败。阈值写反（先 7.0 后 8.5）时，高分会被中档吃掉。

1. **K12.1** 若把上面的 `elif` 全部改成 `if`，`sample` 里哪些记录会被重复计数？先分析再决定要不要运行验证。
2. **K12.2** 只把阈值改成“先 `>= 7.0` 再 `elif >= 8.5`”，预测 high/mid/low，说明高分记录实际进了哪一档。

**作答：** 重复计数：____；调换顺序后的档位：____。


In [ ]:
# Write your predictions, reasoning, or solution here.
# Keep deliberately faulty snippets in Markdown; run your corrected code here.


## 综合练习：独立完成组合筛选与分档

使用同一份 [mobile_game_info.csv](mobile_game_info.csv)。统计全部原始数据记录，标题重复也照常计入；不去重、不按下载量加权。原始采集时间未随文件给出，不要把结果说成当前市场最新水平。

筛选规则必须能用一句话说清。数字碰巧正确但口径说反，不算掌握。

### P1.1　读取与确认列

独立写出读取代码，保留 `rows`、`header`、`body`。打印表头、前三条数据、`len(body)`。根据表头说明 score、game_type、isAndroid 的下标，以及平台字段的真实类型。


In [ ]:
# P1.1: Read and inspect the complete file here.


### P1.2　组合条件平均分

留下 **非安卓，且类型为角色扮演或策略** 的记录。用 `for` + `if` 收集或当场累加，不要使用列表推导式。保留留下的行数、评分总和与平均分。不要把行数或最终结果写死在代码里。

**作答：** 筛选规则（留下谁 / 扔掉谁）：____；条件表达式：____；分母是哪一批记录：____。


In [ ]:
# P1.2: Filter with the combined condition and calculate the subset mean here.


### P1.3　高中低分档与守恒

对 **全部** `body` 记录，按评分分档：`>= 8.5` 为高分，`elif >= 7.0` 为中分，其余为低分。打印三档计数与合计，并确认合计等于 `len(body)`。

从 `body` 中另取连续两条或三条，手算它们各属哪一档，再与同一子集的程序计数核对。不要把小样本档位计数拿去和全表三档数字比较。

重启内核，从头运行已完成的代码。故障片段留在文字中，正常运行单元格只放修正代码。

**作答：** 全表高/中/低/合计：____；守恒是否成立：____；小样本核对：____；重启后的验证情况：____。


In [ ]:
# P1.3: Band all records and record the conservation check here.


## 本章总结

1. 比较运算得到 `True` 或 `False`；`==` 不是 `=`。
2. 文件里的 `"TRUE"` 是字符串，必须按文本比较，不能当成布尔 `True`。
3. `for` 取出每一行，`if` 决定这一行去留；缩进决定哪些语句重复执行。
4. `and` 要求同时成立，`or` 要求至少一侧成立，并且不互斥；组合条件用括号写清口径。
5. 子集平均分的分母只数被留下的记录；评分先 `float` 再比较、再累加。
6. `if` / `elif` / `else` 分档必须互斥，三类计数之和等于参与分档的记录数。

保留读取代码、组合筛选与分档计数。下次可以用字典统计“每个类型多少条”，而不再写一长串 `if`。

课后请打开 [chapter04_条件与判断_课后练习.ipynb](chapter04_条件与判断_课后练习.ipynb)，使用 [googleplaystore.csv](googleplaystore.csv) 独立完成 P1、P2（P3 选做）。两份数据的列名和列数不同，不能把课堂文件的下标直接套到课后文件上。
